File: src/backend/api/routes/asistencia_sesion_service.py
import os, httpx, json
from dotenv import load_dotenv
from fastapi import APIRouter
from fastapi.encoders import jsonable_encoder

from api.models.asistencia_sesion import AsistenciaSesionIn

load_dotenv(dotenv_path=os.path.join(os.path.dirname(__file__), "../../.env"))

# Cliente global para mantener cookies entre peticiones
session = httpx.AsyncClient(
    timeout=20.0,
    follow_redirects=True,
)

BASE_URL = os.getenv("URL_ASISTENCIA_SESIONES")
HEADERS = {
    "Accept": "application/json",
    "User-Agent": "FastAPI-Client",
}

app = APIRouter(prefix="/asistencias", tags=["asistencias"])


async def ensure_cookies():
    """
    Hace una petición inicial si aún no se han guardado cookies en la sesión.
    """
    if not session.cookies:
        resp = await session.get(BASE_URL, headers=HEADERS)
        resp.raise_for_status()


@app.get("/")
async def get_todas_asistencias():
    """Obtener todas las asistencias"""
    await ensure_cookies()
    try:
        resp = await session.get(BASE_URL, headers=HEADERS)
        resp.raise_for_status()
        return resp.json()
    except httpx.TimeoutException:
        return {"error": "Timeout: ORDS no respondió"}
    except httpx.HTTPStatusError as exc:
        return {
            "error": f"HTTP {exc.response.status_code}",
            "detalle": exc.response.text,
        }
    except Exception as exc:
        return {"error": str(exc)}


@app.get("/{id}")
async def get_asistencia(id: int):
    """Obtener una asistencia específica por ID"""
    await ensure_cookies()
    url = f"{BASE_URL}{id}"
    try:
        resp = await session.get(url, headers=HEADERS)
        if resp.status_code in (200, 204):
            return resp.json()
        if resp.status_code in (400, 404):
            return {"error": f"Asistencia con ID {id} no encontrada"}
        return {"status": resp.status_code, "detalle": resp.text}
    except httpx.TimeoutException:
        return {"error": "Timeout: ORDS no respondió"}
    except Exception as exc:
        return {"error": str(exc)}


@app.get("/persona/{id_persona}")
async def get_asistencias_por_persona(id_persona: int):
    """Obtener todas las asistencias de una persona (histórico)"""
    await ensure_cookies()
    try:
        query = {"id_persona": str(id_persona)}
        url = f"{BASE_URL}?q={json.dumps(query)}"
        resp = await session.get(url, headers=HEADERS)
        resp.raise_for_status()
        return resp.json()
    except httpx.TimeoutException:
        return {"error": "Timeout: ORDS no respondió"}
    except httpx.HTTPStatusError as exc:
        return {
            "error": f"HTTP {exc.response.status_code}",
            "detalle": exc.response.text,
        }
    except Exception as exc:
        return {"error": str(exc)}


@app.get("/sesion/{id_sesion}")
async def get_asistencias_por_sesion(id_sesion: int):
    """Obtener todas las asistencias de una sesión específica"""
    await ensure_cookies()
    try:
        query = {"id_sesiones": str(id_sesion)}
        url = f"{BASE_URL}?q={json.dumps(query)}"
        resp = await session.get(url, headers=HEADERS)
        resp.raise_for_status()
        return resp.json()
    except httpx.TimeoutException:
        return {"error": "Timeout: ORDS no respondió"}
    except httpx.HTTPStatusError as exc:
        return {
            "error": f"HTTP {exc.response.status_code}",
            "detalle": exc.response.text,
        }
    except Exception as exc:
        return {"error": str(exc)}

@app.post("/")
async def create_asistencia(asistencia_sesion: AsistenciaSesionIn):
    """Crear una nueva asistencia (registro de asistencia a una sesión)"""
    url = f"{BASE_URL}"
    await ensure_cookies()
    try:
        resp = await session.post(
            url,
            headers={**HEADERS, "Content-Type": "application/json"},
            json=jsonable_encoder(asistencia_sesion),
        )
        resp.raise_for_status()
        if resp.status_code in (200, 201, 204):
            return {"message": "Asistencia registrada correctamente"}
        return {"status": resp.status_code, "detalle": resp.text}
    except httpx.TimeoutException:
        return {"error": "Timeout: ORDS no respondió"}
    except Exception as exc:
        return {"error": str(exc)}


@app.delete("/{id}")
async def delete_asistencia(id: int):
    """Eliminar una asistencia por ID"""
    await ensure_cookies()
    url = f"{BASE_URL}{id}"
    try:
        resp = await session.delete(url, headers=HEADERS)
        if resp.status_code in (200, 204):
            return {"message": f"Asistencia con ID {id} eliminada correctamente"}
        if resp.status_code in (400, 404):
            return {"error": f"Asistencia con ID {id} no encontrada"}
        return {"status": resp.status_code, "detalle": resp.text}
    except httpx.TimeoutException:
        return {"error": "Timeout: ORDS no respondió"}
    except Exception as exc:
        return {"error": str(exc)}


@app.put("/{id}")
async def update_asistencia(id: int, asistencia_sesion: AsistenciaSesionIn):
    """Actualizar una asistencia existente"""
    await ensure_cookies()
    url = f"{BASE_URL}{id}"
    try:
        resp = await session.put(
            url,
            headers={**HEADERS, "Content-Type": "application/json"},
            json=asistencia_sesion.dict(),
        )
        if resp.status_code in (200, 204):
            return {"message": f"Asistencia con ID {id} actualizada correctamente"}
        return {"status": resp.status_code, "detalle": resp.text}
    except httpx.TimeoutException:
        return {"error": "Timeout: ORDS no respondió"}
    except Exception as exc:
        return {"error": str(exc)}

File: src/frontend/lib/api/routes/asistencia_sesion_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class AsistenciaService {
  final String baseUrl;
  final Map<String, String> headers = {
    'Accept': 'application/json',
    'User-Agent': 'Flutter-Client',
  };

  // Simula las cookies de sesión
  Map<String, String> cookies = {};

  AsistenciaService(this.baseUrl);

  // Asegura que haya cookies antes de hacer peticiones
  Future<void> _ensureCookies() async {
    if (cookies.isEmpty) {
      final response = await http.get(Uri.parse(baseUrl), headers: headers);
      if (response.statusCode == 200) {
        _updateCookies(response);
      } else {
        throw Exception('No se pudieron obtener cookies');
      }
    }
  }

  // Actualiza cookies de la sesión
  void _updateCookies(http.Response response) {
    final rawCookies = response.headers['set-cookie'];
    if (rawCookies != null) {
      final cookie = rawCookies.split(';')[0];
      cookies['cookie'] = cookie;
      headers['cookie'] = cookie;
    }
  }

  // GET /asistencias/
  Future<List<dynamic>> getTodasAsistencias() async {
    await _ensureCookies();
    final response = await http.get(Uri.parse(baseUrl), headers: headers);

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Error al obtener asistencias: ${response.statusCode}');
    }
  }

  // GET /asistencias/{id}
  Future<Map<String, dynamic>> getAsistencia(int id) async {
    await _ensureCookies();
    final response = await http.get(Uri.parse('$baseUrl$id'), headers: headers);

    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else if (response.statusCode == 404) {
      throw Exception('Asistencia con ID $id no encontrada');
    } else {
      throw Exception('Error al obtener asistencia: ${response.statusCode}');
    }
  }

  // GET /asistencias/persona/{id_persona}
  Future<List<dynamic>> getAsistenciasPorPersona(int idPersona) async {
    await _ensureCookies();
    final query = jsonEncode({'id_persona': idPersona});
    final url = Uri.parse('$baseUrl?q=$query');

    final response = await http.get(url, headers: headers);
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Error al obtener asistencias por persona: ${response.statusCode}');
    }
  }

  // GET /asistencias/sesion/{id_sesion}
  Future<List<dynamic>> getAsistenciasPorSesion(int idSesion) async {
    await _ensureCookies();
    final query = jsonEncode({'id_sesiones': idSesion});
    final url = Uri.parse('$baseUrl?q=$query');

    final response = await http.get(url, headers: headers);
    if (response.statusCode == 200) {
      return jsonDecode(response.body);
    } else {
      throw Exception('Error al obtener asistencias por sesión: ${response.statusCode}');
    }
  }

  // POST /asistencias/
  Future<void> createAsistencia(Map<String, dynamic> asistencia) async {
    await _ensureCookies();
    final response = await http.post(
      Uri.parse(baseUrl),
      headers: {...headers, 'Content-Type': 'application/json'},
      body: jsonEncode(asistencia),
    );

    if (![200, 201, 204].contains(response.statusCode)) {
      throw Exception('Error al crear asistencia: ${response.statusCode}');
    }
  }

  // PUT /asistencias/{id}
  Future<void> updateAsistencia(int id, Map<String, dynamic> asistencia) async {
    await _ensureCookies();
    final response = await http.put(
      Uri.parse('$baseUrl$id'),
      headers: {...headers, 'Content-Type': 'application/json'},
      body: jsonEncode(asistencia),
    );

    if (![200, 204].contains(response.statusCode)) {
      throw Exception('Error al actualizar asistencia: ${response.statusCode}');
    }
  }

  // DELETE /asistencias/{id}
  Future<void> deleteAsistencia(int id) async {
    await _ensureCookies();
    final response = await http.delete(Uri.parse('$baseUrl$id'), headers: headers);

    if (![200, 204].contains(response.statusCode)) {
      throw Exception('Error al eliminar asistencia: ${response.statusCode}');
    }
  }
}
